File: 1530. Number of Good Leaf Nodes Pairs.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
 func countPairs(root *TreeNode, distance int) int {
    path := []byte{}
    paths := []string{}
    var findPaths func(*TreeNode)
    findPaths = func (current *TreeNode) {
        if current == nil {
            return
        }
        if current.Left == nil && current.Right == nil {
            paths = append(paths, string(path))
        }
        path = append(path, 'L')
        findPaths(current.Left)
        path = path[:len(path)-1]
        path = append(path, 'R')
        findPaths(current.Right)
        path = path[:len(path)-1]
    }
    findPaths(root)
    count := 0
    for i := 0; i < len(paths); i++ {
        for j := i+1; j < len(paths); j++ {
            if check(paths[i], paths[j]) <= distance {
                count++
            }
        }
    }
    return count
}
func check(s, t string) int {
    i, j := 0, 0
    for i < len(s) && j < len(t) {
        if s[i] != t[j] {
            break
        }
        i++
        j++
    }
    return len(s)-i + len(t)-j
}

File: java/1530. Number of Good Leaf Nodes Pairs.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public int countPairs(TreeNode root, int distance) {
        return helper(root, distance)[11];
    }
    private int[] helper(TreeNode root, int distance) {
        if (root == null) return new int[12];
        if (root.left == null && root.right == null) {
            int[] current = new int[12];
            current[0] = 1;
            return current;
        }
        int[] left = helper(root.left, distance);
        int[] right = helper(root.right, distance);
        int[] current = new int[12];
        for (int i = 0; i < 10; i++) {
            current[i+1] += left[i] + right[i];
        }
        current[11] += left[11] + right[11];
        for (int d1 = 0; d1 <= distance; d1++) {
            for (int d2 = 0; d2 <= distance; d2++) {
                if (2 + d1 + d2 <= distance) {
                    current[11] += left[d1] * right[d2];
                }
            }
        }
        return current;
    }
}